Просьба проверить в чём ошибка разработки

Froy-Z opened the pull request:
1. Учтены ошибки первого задания.
2. Проведена декомпозиция
3. Написаны JavaDoc комментарий
4. Настроен docker-compose.yml и Dockerfile
5. Переделана логика сохранения и загрузки данных из БД.
6. Настроены миграции.

Проект не готов к работе из-за непонимания как обновить sequences.

Также не созданы автотесты.

Предпринято множество попыток исправить логику присвоения id_seq.

Не надеюсь пройти на следующий этап, но уже дело принципа понять, что не так с настройкой последовательностей...

Froy-Z committed:
1. Учтены прошлые замечания
2. Проведена декомпозия
3. Написаны JavaDoc комментарий
4. Настроен docker-compose.yml и Dockerfile
5. Переделана логика сохранения и загрузки данных из БД.
6. Настроены миграции.

Проект не готов к работе из-за непонимания как обновить sequences.

Также не созданы автотесты.

Предпринято множество попыток исправить логику присвоения id_seq

YlabCourse commented on src/main/resources/liquibase/changelog-tables.xml:
```diff
@@ -0,0 +1,144 @@
+    <changeSet id="7" author="Froy-Z">
+        <sql>
+        <![CDATA[
+            CREATE TABLE ylab_schema.players
```
нужно было создавать с использованием xml-тегов. Посмотри примеры на https://docs.liquibase.com/change-types/create-table.html. Это важно, чтобы при переходе на другой тип базу данных не переписывать все скрипты

Froy-Z replied:
Создал последовательности с использованием xml-тегов.

YlabCourse commented on src/main/resources/liquibase/changelog-tables.xml:
```diff
@@ -0,0 +1,144 @@
+        <![CDATA[
+            CREATE TABLE ylab_schema.accounts
+            (
+                id        SERIAL PRIMARY KEY,
```
sequence создается примерно вот так createSequence schemaName="test_schema"  sequenceName="test_seq" startValue="1" incrementBy="1" minValue="1" maxValue="90000" cycle="false"

И назначается колонке: column name="id" type="BIGINT" defaultValueComputed="nextval('test_schema.test_seq')"

Froy-Z replied:
Реализовано.

YlabCourse commented on src/main/java/ru/ylab/models/Transaction.java:
```diff
@@ -1,41 +1,29 @@
+    private Long accountId;
+    private String transactionId;
+    private Timestamp executeTime;
+    private double amount;
```
Использование Double для денег плохая практика

Froy-Z replied:
Изучен вопрос, переделан под BigDecimal.

YlabCourse commented on src/main/java/ru/ylab/out/jdbc/JdbcTransaction.java:
```diff
@@ -0,0 +1,88 @@
+    @Override
+    public void save(Transaction transaction) {
+        String sql = "INSERT INTO ylab_schema.transactions (player_id, account_id, transaction_uuid, execution_time, amount, type) " +
+                "VALUES (nextval('transactions_id_seq'), ?, ?, gen_random_uuid(), ?, ?)";
```
gen_random_uuid() ?? Уникальный идентификатор транзакции должен предоставить пользователь

Froy-Z replied:
Исправлено, уникальный идентификатор предоставляет сервис транзакций.

YlabCourse commented on src/main/java/ru/ylab/services/AccountService.java:
```diff
@@ -1,137 +1,148 @@
-                            Math.round((amount - account.getBalance()) * 100.0) / 100.0)
-            );
+    public Logging performCreditTransaction(Player player, Account account, double amount) throws UnknownError {
+        try (Connection connection = DatabaseConnector.getConnection()){
```
сервис не общается с бд напрямую, только через репозиторий

Froy-Z replied:
Исправлено, всё взаимодействие через репозитории.

YlabCourse commented on src/main/java/ru/ylab/out/DatabaseConnector.java:
```diff
@@ -0,0 +1,50 @@
+
+            // миграция создания схемы для служебной информации
+            Liquibase liquibase = new Liquibase("liquibase/changelog.xml", new ClassLoaderResourceAccessor(), database);
+            liquibase.update();
```
Служебные таблицы Liquibase не должны хранится в схеме public

Froy-Z replied:
Исправлено, служебные таблицы в отдельной схеме.

YlabCourse commented on src/main/java/ru/ylab/out/DatabaseConnector.java:
```diff
@@ -0,0 +1,50 @@
+ */
+@NoArgsConstructor
+public class DatabaseConnector {
+    private static final String URL = "jdbc:postgresql://localhost:2345/ylab_db";
```
Параметры подключения к базе данных должны задаваться в конфигурационном файле

Froy-Z replied:
Исправлено, подключение к БД из конфиг. файла .properties

YlabCourse commented on src/main/java/ru/ylab/out/DatabaseConnector.java:
```diff
@@ -0,0 +1,50 @@
+            Database database = DatabaseFactory.getInstance().findCorrectDatabaseImplementation(new JdbcConnection(connection));
+
+            // миграция создания схемы для служебной информации
+            Liquibase liquibase = new Liquibase("liquibase/changelog.xml", new ClassLoaderResourceAccessor(), database);
```
Параметры Liquibase  должны задаваться в конфигурационном файле

Froy-Z replied:
Исправлено, путь к конфиг. файлам миграции указан через .properties

Froy-Z committed:
1. Созданы последовательности sequences.
2. Настроены конфиг. файлы для БД и миграций.
3. Настроены миграции Liquibase;
4. Создание схемы для служебных таблиц.
5. Сохранение и выгрузка в базу данных PostgreSQL(БД поднимается в docker-контейнере) через репозитории;
6. Настроен пулл соединений (Hikari).

Реализовано:
- Регистрация игрока;
- Авторизация по логину-паролю;
- Заведение кошелька(счёта) игроку;
- Пополнение кошелька;
- Списание с кошелька;
- Просмотр аудита всех операций игрока;
- Просмотр истории денежных потоков игрока;
- Деавторизация;

Проведена декомпозиция, улучшена читаемость кода, проект в рабочем состоянии.

Оставшиеся задачи на бранч:
- тест-контейнеры;
- javaDoc для классов и методов.
